### ISSUE TITLE
Keycloak 22.0.4 crashes on file login-status-iframe.html on old browser versions ( Google Chrome below 80 )

### ISSUE THREAD

**Issue by agapovx** [keycloak/keycloak#23902]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

authentication

### Describe the bug

On old browser versions ( for example, _Google Chrome version below 80_ ) the file `login-status-iframe.html` crashes because there's an javascript operator `??` is not supported on this version ( 79 and lower );

Here ie the console error:

`Uncaught SyntaxError: Unexpected token '?'`

This error happened on the next code `return cookies.get(name) ?? null;` (`login-status-iframe.html` 101 line of code );

It will be fine if this error will be fixed in the nearest patch ( 22.0.5 );

Similar problem in this issue ( problem connected with old browser versions ) - https://github.com/keycloak/keycloak/issues/23743

### Version

22.0.4

### Expected behavior

Keycloak run without runtime javascript error's

### Actual behavior

Keycloak crashes on old browser versions;

### How to Reproduce?

On the frontend app side for `KeycloakInitOptions` you need to set `checkLoginIframe` with value `true`;

Try to connect to the client on old browser version ( for example Google Chrome version 79 );

### Anything else?

I have solution to fix it:

```
<!doctype html>
<html>
<head>
    <meta charset="utf-8">
</head>
<body>
<script type="module">
    window.addEventListener("message", onMessage);

    async function onMessage(event) {
        // Filter out any events that do not match the expected format of a 2-part string split by a space.
        if (typeof event.data !== "string") {
            return;
        }

        const data = event.data.split(" ");

        if (data.length !== 2) {
            return;
        }

        // Extract data from event and verify status of session.
        const [clientId, sessionState] = data;
        const sessionStatus = await checkState(clientId, event.origin, sessionState);

        // Signal session status to the page embedding this iframe.
        event.source.postMessage(sessionStatus, event.origin);
    }

    let init;

    async function checkState(clientId, origin, sessionState) {
        // Check if the browser has granted us access to 3rd-party storage (such as cookies).
        const hasAccess = await hasStorageAccess();

        // If we don't have access, signal an error.
        // As we cannot read cookies, we cannot verify the session state.
        if (!hasAccess) {
            return "error";
        }

        // If not initialized, verify this client is allowed access with a call to the server.
        if (!init) {
            const url = new URL(`${location.origin}${location.pathname}/init`);

            url.searchParams.set("client_id", clientId);
            url.searchParams.set("origin", origin);

            const response = await fetch(url);

            if (!response.ok) {
                return "error";
            }

            init = { clientId, origin };
        }

        const cookie = getSessionCookie();

        // Signal a change in state if there is no cookie, and the session state is not empty.
        if (!cookie) {
            return sessionState !== "" ? "changed" : "unchanged";
        }

        // If the client and origin from the event match the verified ones from the server, signal if the cookie has changed.
        if (clientId === init.clientId && origin === init.origin) {
            const [, , cookieSessionState] = cookie.split("/");
            return sessionState === cookieSessionState ? "unchanged" : "changed";
        }

        // Otherwise, if there is no match, then signal an error.
        return "error";
    }

    async function hasStorageAccess() {
        if (!("hasStorageAccess" in document)) {
            return true;
        }

        return document.hasStorageAccess();
    }

    function getSessionCookie() {
        const cookie = getCookieByName("KEYCLOAK_SESSION");

        if (cookie !== null) {
            return cookie;
        }

        return getCookieByName("KEYCLOAK_SESSION_LEGACY");
    }

    function getCookieByName(name) {
        const cookies = new Map();

        for (const cookie of document.cookie.split(";")) {
            const [key, value] = cookie.split("=").map((value) => value.trim());
            cookies.set(key, value);
        }

        const cookieByName = cookies.get(name);

        return (cookieByName !== null && cookieByName !== undefined) ? cookieByName : null;
    }
</script>
</body>
</html>

```

**Comment by agapovx** [comment]
I will make PR to fix this problem;

**Comment by khidirbekov** [comment]
+1, I have the same problem

**Comment by dzvyagin** [comment]
Yeah, I've encountered this issue too and it would be really cool if it gets resolved as I had to set checkLoginIframe to false to resolve it

**Comment by jonkoops** [comment]
The [nullish coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) that is causing this issue is part of the ES2020 version of the ECMAScript specification. This version of the specification in the minimum version we require at this time to run Keycloak. 

The nullish coalescing operator has been in Chrome since version 80, which released in February 2020, almost four years ago. This version of Chrome is no longer supported by us, nor by Google itself, and has several known critical security issues, including remote code execution vulnerabilities. If you still have users that are making use of this browser they are at non-insignificant risk of attack.

As mentioned, because the minimum ECMAScript version we require is at least ES2020, and only unmaintained browser versions are affected we will not be fixing this particular issue. Instead we recommend that users upgrade their browser to get the best, and more importantly secure experience.

> Similar problem in this issue ( problem connected with old browser versions ) - https://github.com/keycloak/keycloak/issues/23743

I think most folks here are likely seeing this syntax error, as this caused by [top level await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#top_level_await), which is a ES2023 feature. This should not have been introduced, and a PR to fix this is already open under #23793. We will resolve this issue in the next patch release of Keycloak (22.0.5).

**Comment by agapovx** [comment]
This is not good news for me.
I understand your's position ( about support at least ES2020 ) but some of user's can't update browser version by some restrictions ( this is not good, but we can't do anything at this situation ); Count of this user's is over 1,000;
But we need to update keycloak's version because of some bug-fixes ( in last version's ).

And finally we have situation: update keycloak version by fixing some of important bugs BUT we have another bugs after update ( one about my issue and another one about top level await );

Maybe you will do some exception for my issue and approve PR with 2 lines of code? This is very small PR which does not change anything by app behavior but fix some crashes in old browser versions;

And I think that not only me have this situation because keycloak - it is a good one opensource application;

@jonkoops 
I hope for your's approval.
Thank you.

**Comment by jonkoops** [comment]
I am really sorry @agapovx, but we have to draw the line somewhere. I would recommend you advise your users to upgrade their browser instead.
